src/GitHub.App/Services/TeamExplorerContext.cs

```diff
@@ -0,0 +1,176 @@
+﻿using System;
+using System.Linq;
+using System.ComponentModel;
+using System.ComponentModel.Composition;
+using System.Collections.Generic;
+using Microsoft.VisualStudio.Shell;
+using GitHub.Models;
+using GitHub.Logging;
+using GitHub.Primitives;
+using Serilog;
+using EnvDTE;
+
+namespace GitHub.Services
+{
+    /// <summary>
+    /// This service uses reflection to access the IGitExt from Visual Studio 2015 and 2017.
+    /// </summary>
+    [Export(typeof(ITeamExplorerContext))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class TeamExplorerContext : ITeamExplorerContext
+    {
+        const string GitExtTypeName = "Microsoft.VisualStudio.TeamFoundation.Git.Extensibility.IGitExt, Microsoft.TeamFoundation.Git.Provider";
+
+        readonly ILogger log;
+        readonly DTE dte;
+
+        string solutionPath;
+        string repositoryPath;
+        string branchName;
+        string headSha;
+        bool testing;
+
+        [ImportingConstructor]
+        public TeamExplorerContext([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
+            : this(serviceProvider, LogManager.ForContext<TeamExplorerContext>(), null, false)
+        {
+        }
+
+        public TeamExplorerContext([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider, ILogger log,
+            Type gitExtType, bool testing)
+        {
+            this.log = log;
+            this.testing = testing;
+
+            gitExtType = gitExtType ?? FindGitExtType();
```

Comment:
The problem is there are two versions of this assembly with no binding redirections between them. I wanted to avoid dependencies on GitHub.TeamFoundation.14/15.

Reply:
Ahhh, ok yeah that makes more sense then. Maybe add a comment to this effect to the source.